src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -2,32 +2,193 @@
+    /// into the output.   This is not likely to be a good idea as it will be very verbose, but is useful to quickly
+    /// discover what is available.
+    /// 
+    /// See the DiagnosticSourceEventSourceBridgeTest.cs for more explicit examples of using this bridge.  
     /// </summary>
     [EventSource(Name = "Microsoft-Diagnostics-DiagnosticSource")]
     internal class DiagnosticSourceEventSource : EventSource
     {
         public static DiagnosticSourceEventSource Logger = new DiagnosticSourceEventSource();
 
+        class Keywords
+        {
+            /// <summary>
+            /// Indicates diagnostics messages from DiagnosticSourceEventSource should be included. 
+            /// </summary>
+            public const EventKeywords Messages = (EventKeywords)1;
+            /// <summary>
+            /// Indicates that all events from all diagnostic sources should be forwarded to the EventSource using the 'Event' event.  
+            /// </summary>
+            public const EventKeywords Events = (EventKeywords)2;
+        };
+
+        /// <summary>
+        /// Used to send ad-hoc diagnostics to humans.   
+        /// </summary>
+        [Event(1, Keywords = Keywords.Messages)]
+        public void Message(string Message)
+        {
+            WriteEvent(1, Message);
+        }
+
+#if !NO_EVENTSOURCE_COMPLEX_TYPE_SUPPORT
+        /// <summary>
+        /// Events from DiagnosticSource can be forwarded to EventSource using this event.  
+        /// </summary>
+        [Event(2, Keywords = Keywords.Events)]
+        private void Event(string SourceName, string EventName, IEnumerable<KeyValuePair<string, string>> Arguments)
+        {
+            WriteEvent(2, SourceName, EventName, Arguments);
+        }
+#endif 
+        /// <summary>
+        /// This is only used on V4.5 systems that dont have the ability to log KeyValuePairs directly.
+        /// It will eventually go away, but we should alwasy reserve the ID for this.    
+        /// </summary>
+        [Event(3)]
+        private void EventJSON(string SourceName, string EventName, string ArgmentsJSON)
+        {
+            WriteEvent(3, SourceName, EventName, ArgmentsJSON);
+        }
+
+        #region private
+
+#if NO_EVENTSOURCE_COMPLEX_TYPE_SUPPORT
         /// <summary>
-        /// On every command (which the debugger can force by turning on this EventSource with ETW)
-        /// call a function that the debugger can hook to do an arbitrary func evaluation.  
+        /// Converts a keyvalue bag to JSON.  Only used on V4.5 EventSources.  
         /// </summary>
-        /// <param name="args"></param>
-        protected override void OnEventCommand(EventCommandEventArgs args)
+        private static string ToJSON(List<KeyValuePair<string, string>> keyValues)
         {
+            StringBuilder sb = new StringBuilder();
+            sb.AppendLine("{");
```

Comment:
It turns out that AppendLine does not have a character overload, just a string one.   It is also hopefully a non-common code path (only used in V4.5 case), so I did not try to optimize it (e.g. reuse the stringbuilder ...)